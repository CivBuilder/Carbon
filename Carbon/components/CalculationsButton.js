import { View, Text, StyleSheet, Pressable } from 'react-native'
import React from 'react'


const CalculationsButton = ({ onPress }) => {
    return (
        <View >
            <Pressable style={styles.button} onPress={onPress} testID="calculation-details">
                <Text style={styles.buttonText}>Calculation Details</Text>
            </Pressable >
        </View>

    )
}

const styles = StyleSheet.create({
    button: {
        width: 200,
        height: 40,
        backgroundColor: '#74C69D',
        borderRadius: 20,
        alignItems: 'center',
        justifyContent: 'center',
        marginTop: 10,
        marginBottom: 30,
    },
    buttonText: {
        color: 'white',
        fontFamily: 'sans-serif',
    },
})
export default CalculationsButton;