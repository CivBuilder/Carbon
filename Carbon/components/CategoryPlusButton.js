import { StyleSheet, Text, TextInput, View, Pressable, Modal } from 'react-native'
import {React, useState} from 'react'
import SaveEmissions from './SaveEmissions';

const CategoryPlusButton = () => {
  const [modalVisible, setModalVisible] = useState(false);
  const [foodVisible, setFoodVisible] = useState(false);
  const [transportationVisible, setTransportationVisible] = useState(false);
  const [recyclingVisible, setRecyclingVisible] = useState(false);

  
  //TODO: Sanitize and save input to database
  
  const saveFoodLog = (log) => {
    console.log('saving food ' + log);
  }
  const saveTransportationLog = (log) => {
    console.log('saving transportation ' + log);
  }

  const saveRecyclingLog = (log) => {
    console.log('saving recycling ' + log);
  }
  return (
    <View>
        <Modal style={styles.modal}
        animationType="slide"
        transparent={true}
        backdropOpacity={0}
        visible={modalVisible}
        onRequestClose={() => {
          setModalVisible(!modalVisible);
        }}>
        <View style={styles.centeredView}>
          <View style={styles.modalView}>

            <Text style={styles.modalText}>Select the category to log</Text>
            <Pressable onPress={() => setFoodVisible(true)}>
              <Text style={styles.text}>Food</Text>
            </Pressable>

            <Pressable onPress={() => setTransportationVisible(true)}>
              <Text style={styles.text}>Transportation</Text>
            </Pressable>

            <Pressable onPress={() => setRecyclingVisible(true)}>
              <Text style={styles.text}>Recycling</Text>
            </Pressable>

            <Pressable
              style={[styles.button]}
              onPress={() => setModalVisible(!modalVisible)}>
              <Text style={styles.text}>close</Text>
            </Pressable>

          </View>
        </View>
        </Modal>

      <SaveEmissions visible={foodVisible}
                     onRequestClose={() => {
                      setFoodVisible(!foodVisible)
                    }}  
                    title={"How much meat did you consume today?"}
                    saveData={(log) => saveFoodLog(log)}
                    />
        <SaveEmissions visible={transportationVisible}
                     onRequestClose={() => {
                      setTransportationVisible(!transportationVisible)
                    }}  
                    title={"How many miles did you drive today?"}
                    saveData={(log) => saveTransportationLog(log)}
                    />
        <SaveEmissions visible={recyclingVisible}
                     onRequestClose={() => {
                      setRecyclingVisible(!recyclingVisible)
                    }}  
                    title={"How much did you recycle today?"}
                    saveData={(log) => saveRecyclingLog(log)}
                    />

        <Pressable style={styles.button} onPress={() => setModalVisible(true)}>
        <Text style={styles.text}>Add</Text>
        </Pressable>
    </View>
    
  )
}

export default CategoryPlusButton

const styles = StyleSheet.create({
    centeredView: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        marginTop: 22,
        justifyContent: 'flex-end',
        marginBottom: 48 ,
      },

    modalView: {
        width: "100%",
        height: "33%",
        borderTopStartRadius: 12,
        borderTopEndRadius: 12,
        alignItems: "center",
        justifyContent: "center",
        backgroundColor: '#465966',
        shadowOpacity: 0.25,
        shadowRadius: 4,
        elevation: 5,
      },
    modal: {
      backgroundColor: '#465966',
    },  
    text: {
      textAlign: 'center',
    },
    button: {
        borderRadius: 4,
        borderWidth: 2,
        borderColor: 'black',
        padding: 10,
        margin: 10,
        elevation: 2,
        width: 120,
      },

      modalText: {
        marginBottom: 15,
        textAlign: 'center',
      },
})