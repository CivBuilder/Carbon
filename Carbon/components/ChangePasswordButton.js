import { View, Text, StyleSheet, Pressable } from 'react-native'
import React from 'react'


const ChangePasswordButton = ({disabled, onPress }) => {
    return (
        <View >
            {disabled ? (
                <Pressable style={{...styles.button, backgroundColor: '#C2C2C2'}} onPress={onPress} testID="changePasswordButton" disabled={disabled}>
                    <Text style={styles.buttonText}>Change Password</Text>
                </Pressable >
            ) : (
                <Pressable style={styles.button} onPress={onPress} testID="changePasswordButton">
                    <Text style={styles.buttonText}>Change Password</Text>
                </Pressable >
            )}
        </View>

    )
}

const styles = StyleSheet.create({
    button: {
        width: 200,
        height: 40,
        backgroundColor: '#74C69D',
        borderRadius: 20,
        alignItems: 'center',
        justifyContent: 'center',
        marginTop: 10,
        marginBottom: 30,
    },
    buttonText: {
        color: 'white',
        fontSize: 16,
        fontFamily: 'sans-serif',
    },
})

export default ChangePasswordButton