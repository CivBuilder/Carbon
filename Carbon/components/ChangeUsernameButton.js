import { View, Text, StyleSheet, Pressable } from 'react-native'
import React from 'react'

const ChangeUsernameButton = ({ onPress }) => {
    return (
        <View >
            <Pressable style={styles.button} onPress={onPress} testID="changeUsernameButton">
                <Text style={styles.buttonText}>Change Username</Text>
            </Pressable >
        </View>
    )
}

const styles = StyleSheet.create({
    button: {
        width: 200,
        height: 40,
        backgroundColor: '#74C69D',
        borderRadius: 20,
        alignItems: 'center',
        justifyContent: 'center',
        marginTop: 10,
        marginBottom: 30,
    },
    buttonText: {
        color: 'white',
        fontFamily: 'sans-serif',
    },
})

export default ChangeUsernameButton