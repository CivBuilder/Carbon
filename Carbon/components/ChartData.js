import React, { useState } from 'react';
import { View, Text, StyleSheet, Dimensions } from 'react-native';
import { VictoryPie } from 'victory-native';
import { Colors } from '../colors/Colors';

const windowWidth = Dimensions.get("window").width;
const windowHeight = Dimensions.get("window").height;
const margin = 10;
const chartWidth = windowWidth - (margin * 2);
const chartHeight = 210;

const monthNames = [
    "January", "February", "March", "April", "May", "June", "July",
    "August", "September", "October", "November", "December"
];
const today = new Date();
const currentMonth = monthNames[today.getMonth()];

const data = [
    { x: "Transport", y: Math.round(Math.random() * 10000) },
    { x: "Diet", y: Math.round(Math.random() * 10000) },
    { x: "Home", y: Math.round(Math.random() * 10000) },
    { x: "Stuff", y: Math.round(Math.random() * 10000) },
];

// Ticket C4-19
export const CatgegoryChart = () => {
    const total = data.reduce((acc, datum) => acc + datum.y, 0); //Gets the overall value of all categories

    //Components for the pie chart
    const pieRadius = (windowWidth - (margin * 2)) * 0.4;
    const innerRadius = pieRadius * 0.65;
    const labelRadius = ((pieRadius + innerRadius) / 2);

    // Renders the label for each section of the pie chart
    const getLabel = (datum) => {
        const percent = Math.round((datum.y / total) * 100);
        if (percent > 4) {
            return `${datum.x}\n${Math.round((datum.y / total) * 100)}%`;
        }
        return null;
    };

    const [selectedSlice, setSelectedSlice] = useState(null);

    const handlePress = (event, props) => {
        const selectedDatum = data[props.index];
        setSelectedSlice(props.index);
        // console.log(`Selected slice: ${selectedDatum.x} (${selectedDatum.y})`);
    };

    // Shows the value of the selected section in the middle of the pie chart
    const getSelectedLabel = () => {
        if (selectedSlice !== null) {
            const selectedDatum = data[selectedSlice];
            return `${selectedDatum.x}\n${selectedDatum.y} lbs CO2`;
        }
        return null;
    };

    return (
        <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center', marginVertical: -40 }}>
            <VictoryPie
                data={data}
                colorScale={["#5D5FEF", "#FF8C00", "#3CB371", "#FF69B4"]}
                padAngle={2}
                radius={pieRadius}
                innerRadius={innerRadius}
                cornerRadius={6}
                labelRadius={labelRadius}
                labels={({ datum }) => getLabel(datum)}
                style={{
                    labels: {
                        fill: "white",
                        fontSize: 12,
                        fontWeight: "bold",
                        textAnchor: "middle",
                        verticalAnchor: "middle",
                    },
                }}
                events={[{
                    target: 'data',
                    eventHandlers: {
                        onPressIn: handlePress,
                    },
                },]}
                animate={{ duration: 500, ease: "exp" }}
                selected={[selectedSlice]}
            />
            <View
                style={{
                    justifyContent: 'center',
                    alignItems: 'center',
                    position: 'absolute',
                }}>
                <View style={{ paddingBottom: 5 }}>
                    <Text style={{ textAlign: 'center', fontSize: 20, fontWeight: 'bold', textDecorationLine: 'underline' }}>{currentMonth}</Text>
                </View>

                {selectedSlice == null && ( // A slice has not yet selected
                    <View style={{ }}>
                        <Text style={{ textAlign: 'center' }}>Click a section to learn more</Text>
                    </View>
                )}
                {selectedSlice !== null && ( // A slice has been selected
                    <View style={{ }}>
                        <Text style={{ textAlign: 'center', fontSize: 16 }}>{getSelectedLabel()}</Text>
                    </View>
                )}
            </View>
        </View>
    );
};

//Data for later
/*
    The daily log function, this takens in a data array and maps it into a bar chart based on its values
 */
export function DailyLog ({dataArray}) {

    //Our data for bar styling the bar chart
    const barChartData = {
        labels: ['Transportation', 'Diet', 'Lifestyle', 'Home', 'Overall'],
        datasets:[
            {
            data: dataArray,
            colors: [
                (opacity = 1) =>  Colors.primary.MINT,
                (opacity = 1) =>  Colors.primary.MINT,
                (opacity = 1) =>  Colors.primary.MINT, //all colors will be MINt
                (opacity = 1) =>  Colors.primary.MINT,
                (opacity = 1) =>  Colors.primary.MINT,
            ]
            }
        ]
        };
        //Return the bar chart with appropriate styling.
    return (
        <BarChart
        data ={barChartData}
        chartConfig={styleBar.chartConfig}
        withCustomBarColorFromData={true}
        flatColor={true}
        style={styleBar.chart}
        width = {chartWidth}
        height = {windowHeight/3}
        withVerticalLabels ={true}
        withHorizontalLabels = {true}
        fromZero = {true}
    />
    )
}

const styles = StyleSheet.create({
    chart: {
        borderRadius: 16,
    },
    chartConfig: {
        backgroundColor: Colors.primary.MINT,
        backgroundGradientFrom: Colors.primary.MINT,
        backgroundGradientTo: Colors.primary.MINT,
        decimalPlaces: 2, // optional, defaults to 2dp
        color: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
        labelColor: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
        propsForDots: {
            r: "5",                         // circle size
            strokeWidth: "2",               // circle border size
            stroke: Colors.primary.MINT     // circle border color
        }
    },
});

//Style bar which was created by Miguel
const styleBar = StyleSheet.create({
    container: {
        marginLeft: 100,
        flex: 1,
        justifyContent: 'center', padding: 100,
        paddingTop: 30, backgroundColor: '#ecf0f1',
        },
    chart: {
        borderRadius: 16,
        marginHorizontal: margin/2,
    },
    chartConfig: {

        backgroundColor: '#FFFFFF',
        backgroundGradientFrom: '#FFFFFF',
        backgroundGradientTo: '#FFFFFF',
        decimalPlaces: 2, // optional, defaults to 2dp
        color: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`, //transparent
        labelColor: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,

        propsForDots: {
            r: "5",                         // circle size
            strokeWidth: "2",               // circle border size
            stroke: Colors.primary.MINT     // circle border color
        }
    },
    keyFactors: {
        flex: 1,
        flexDirection: 'column',
        justifyContent: 'space-between',
        alignItems: 'center',
        marginHorizontal: margin,
        marginVertical: margin-5
    },
    category: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        flexWrap: 'nowrap',
        marginVertical: 5
    },
});