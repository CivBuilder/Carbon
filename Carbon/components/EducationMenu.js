import {View, Text, TouchableOpacity, Image, StyleSheet} from 'react-native'
import { Colors } from '../styling/Colors'
import React from 'react'


const EducationMenu = ({title, imageSrc, onPress}) => {
    const test = {
        backgroundColor: '#43b262'
    }
    return(
        <View style={{paddingHorizontal: 10, alignItems: 'center'}}>
            <TouchableOpacity style = {styles.roundButton}  onPress={onPress}>
                <Image style = {styles.categoryIcon} source={imageSrc} />
            </TouchableOpacity>
            <Text style = {styles.imageText}>{title}</Text>
        </View>
   )
}

const styles = StyleSheet.create({
    roundButton: {
        height: 70,
        width: 70,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 100,
        backgroundColor: Colors.primary.MINT_CREAM,
        borderColor: Colors.primary.MINT,
        borderWidth: 3,
    },

    categoryIcon: {
        resizeMode: 'contain',
        height: '80%',
        width: '80%',
        tintColor: Colors.primary.MINT,
    },

    imageText: {
        fontSize: 17,
        fontWeight: 'bold',
        color: 'gray',
        textAlign: 'center'
    }
});


export default EducationMenu
