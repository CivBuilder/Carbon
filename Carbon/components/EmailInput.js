import { View, Text, StyleSheet, TextInput } from 'react-native'
import React from 'react'

const EmailInput = () => {
  return (
    <View>
      <TextInput style={styles.input} placeholder={'Email'} placeholderTextColor="#AFAFAF" />
    </View>
  )
}

const styles = StyleSheet.create({
    input: {
        height: 40,
        backgroundColor: 'white',
        color: '#AFAFAF',
        borderWidth: 2,
        borderColor: '#74C69D',
        borderTopLeftRadius: 20,
        borderTopRightRadius: 20,
        borderBottomLeftRadius: 20,
        borderBottomRightRadius: 20,
        fontSize: 16,
        fontFamily: 'sans-serif',
        marginTop: 10,
        paddingLeft: 15,
    },
})

export default EmailInput