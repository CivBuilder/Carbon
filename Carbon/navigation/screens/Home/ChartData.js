import { View, Text, StyleSheet, Dimensions } from 'react-native';
import { LineChart, BarChart, PieChart, ProgressChart, ContributionGraph, StackedBarChart} from "react-native-chart-kit";
import { ScreenWidth } from 'react-native-elements/dist/helpers';
import { ProgressCircle } from 'react-native-svg-charts';
import { Colors } from '../../../colors/Colors';

const windowWidth = Dimensions.get("window").width;
const windowHeight = Dimensions.get("window").height;
const horizontalMargin = 20;
const chartWidth = windowWidth - horizontalMargin;
const chartHeight = 200;

export const CarbonFootprint = () => {
    const maxFootprint = 100000; //TODO: Change hardcoded data
    const userFootprint = 69420; //TODO: Change hardcoded data
    const percentFootprint = userFootprint / maxFootprint;

    return (
        <ProgressCircle
            style={{ height: 250, marginBottom: -110, marginVertical: 20 }}
            progress={percentFootprint}
            progressColor={Colors.primary.MINT}
            backgroundColor={'rgba(194, 65, 244, 0.1)'}
            strokeWidth={14}
            startAngle={-Math.PI / 2}
            endAngle={Math.PI / 2}
        >
            {/* TODO: Add gradient where it starts mint green at 0 then fades to red at 70% */}
            {/* <LinearGradient id="gradient" x1="0%" y1="0%" x2="100%" y2="0%">
                <Stop offset="0%" stopColor={Colors.primary.MINT} />
                <Stop offset="70%" stopColor={Colors.secondary.CELADON} />
            </LinearGradient> */}
            <View style={{ position: 'absolute', justifyContent: 'center', alignItems: 'center', width: '100%', height: '100%', top: -40,}}>
                <Text style={{ fontSize: userFootprint >= 100000 ? 32 : 40, fontWeight: 'bold' }}>{userFootprint}</Text>
                <Text style={{ fontSize: 20, fontWeight: 'bold' }}>lbs COâ‚‚ </Text>
            </View>
        </ProgressCircle>
    );
};

export const LineChartFootprint = () => {
    const lineChartData = {
        labels: ["Jan", "Feb", "Mar", "Apr", "May", "Jun"],
        datasets: [{
            // TODO: Change hardcode data
            data: [
                Math.random() * 100,
                Math.random() * 100,
                Math.random() * 100,
                Math.random() * 100,
                Math.random() * 100,
                Math.random() * 100
            ]
        }]
    };

    return(
        <LineChart
        data={lineChartData}
        width={chartWidth}
        height={chartHeight}
        //yAxisLabel=""
        yAxisSuffix=" lb"
        yAxisInterval={1}
        chartConfig={styles.chartConfig}
        style={styles.chart}
        bezier
    />
    );
};

export const ProgressRingCategory = () => {
    const progressRingData = {
        labels: ["Transport", "Diet", "Home", "Stuff"],
        data: [0.4, 0.6, 0.8, 0.5] //TODO: Change hardcoded data
    };

    return(
        <ProgressChart
            data={progressRingData}
            width={chartWidth}
            height={chartHeight}
            strokeWidth={12}
            radius={24}
            chartConfig={styles.chartConfig}
            hideLegend={false}
            style={styles.chart}
        />
    );
};

export const DailyLog = () => {
    const barChartData = { 
        labels: ["Group A", "Group B", "Group C"],
        datasets:[
            {
           data: [854, 393, 760]
            }
        ]
        };
    return (
        <BarChart 
        data ={barChartData} 
        chartConfig={styles.chartConfig}
        style={styles.chart}
        width = {chartWidth}
        height = {chartHeight-100}
        withVerticalLabels ={true}
        withHorizontalLabels = {false}
        yAxisLabel="$"
        fromZero = {true}
      />
    )
}
const styles = StyleSheet.create({
    chart: {
        borderRadius: 16,
        marginHorizontal: horizontalMargin/2,
    },
    chartConfig: {
        backgroundColor: Colors.primary.MINT,
        backgroundGradientFrom: Colors.primary.MINT,
        backgroundGradientTo: Colors.primary.MINT,
        decimalPlaces: 2, // optional, defaults to 2dp
        color: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
        labelColor: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
        propsForDots: {
            r: "5",                         // circle size
            strokeWidth: "2",               // circle border size
            stroke: Colors.primary.MINT     // circle border color
        }
    },
});