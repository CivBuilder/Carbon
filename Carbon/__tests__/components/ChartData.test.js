import React from 'react';
import { render } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { CategoryChart, DailyLog } from '../../components/ChartData';
import { getData, fetchData, getLabel, getSelectedLabel } from '../../components/ChartData';

describe('getData', () => {
    const fakeFetch = (data) => {
        return jest.fn().mockImplementation(() => Promise.resolve({
            ok: true,
            json: () => Promise.resolve(data),
        }));
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should return null if no data is fetched', async () => {
        const setLoading = jest.fn();
        const setError = jest.fn();
        global.fetch = fakeFetch([]);
        const result = await getData('2022-01', setLoading, setError);
        expect(result).toBeNull();
        expect(setLoading).toHaveBeenCalledTimes(2);
        expect(setLoading).toHaveBeenNthCalledWith(1, true);
        expect(setLoading).toHaveBeenNthCalledWith(2, false);
        expect(setError).not.toHaveBeenCalled();
    });

    it('should return the emissions data when data is fetched', async () => {
        const setLoading = jest.fn();
        const setError = jest.fn();
        const fetchedData = [
            {
                id: 1,
                user_id: 323,
                date: '2022-01-01',
                transport_emissions: 10,
                lifestyle_emissions: 20,
                home_emissions: 30,
                diet_emissions: 40,
            },
            {
                id: 2,
                user_id: 323,
                date: '2022-01-02',
                transport_emissions: 50,
                lifestyle_emissions: 60,
                home_emissions: 70,
                diet_emissions: 80,
            },
        ];
        global.fetch = fakeFetch(fetchedData);
        const result = await getData('2022-01', setLoading, setError);
        expect(result).toEqual([
            { x: 'Transport', y: 60 },
            { x: 'Lifestyle', y: 80 },
            { x: 'Home', y: 100 },
            { x: 'Diet', y: 120 },
        ]);
        expect(setLoading).toHaveBeenCalledTimes(2);
        expect(setLoading).toHaveBeenNthCalledWith(1, true);
        expect(setLoading).toHaveBeenNthCalledWith(2, false);
        expect(setError).not.toHaveBeenCalled();
    });

    it('should return null and set error when fetch fails', async () => {
        const setLoading = jest.fn();
        const setError = jest.fn();
        const fetchError = new Error('Fetch error');
        global.fetch = jest.fn().mockImplementation(() => Promise.reject(fetchError));
        const result = await getData('2022-01', setLoading, setError);
        expect(result).toBeNull();
        expect(setLoading).toHaveBeenCalledTimes(2);
        expect(setLoading).toHaveBeenNthCalledWith(1, true);
        expect(setLoading).toHaveBeenNthCalledWith(2, false);
        expect(setError).toHaveBeenCalledTimes(1);
        expect(setError).toHaveBeenCalledWith(fetchError);
    });

    // it('should return null and set error when fetch times out', async () => {
    //     expect.assertions(6);

    //     const setLoading = jest.fn();
    //     const setError = jest.fn();

    //     global.fetch = jest.fn().mockImplementation(() => {
    //         return new Promise((resolve, reject) => {
    //         setTimeout(() => {
    //             resolve([]);
    //         }, 30000);
    //         });
    //     });

    //     const resultPromise = getData('2022-01', setLoading, setError);
    //     const result = await Promise.race([resultPromise, new Promise(resolve => setTimeout(resolve, 31000))]);

    //     expect(result).toBeNull();
    //     expect(setLoading).toHaveBeenCalledTimes(2);
    //     expect(setLoading).toHaveBeenNthCalledWith(1, true);
    //     expect(setLoading).toHaveBeenNthCalledWith(2, false);
    //     expect(setError).toHaveBeenCalledTimes(1);
    //     expect(setError.mock.calls[0][0].message).toBe('Network request timed out');
    // }, 30000);
});

describe('fetchData', () => {
    const currentMonth = 'April';
    let setData;
    let setTotal;
    let setLoading;
    let setError;

    beforeEach(() => {
        setData = jest.fn();
        setTotal = jest.fn();
        setLoading = jest.fn();
        setError = jest.fn();
    });

    it('should fetch data and update state when data is retrieved', async () => {
        const emissionsData = [
            { x: "Transport", y: 50 },
            { x: "Lifestyle", y: 30 },
            { x: "Home", y: 20 },
            { x: "Diet", y: 10 },
        ];
        const getTotal = emissionsData.reduce((acc, datum) => acc + datum.y, 0);
        const getData = jest.fn().mockResolvedValue(emissionsData);

        await fetchData(currentMonth, setData, setTotal, setLoading, setError, getData);

        expect(setLoading).toHaveBeenCalledTimes(2);
        expect(setLoading).toHaveBeenCalledWith(true);
        expect(setLoading).toHaveBeenCalledWith(false);
        expect(setData).toHaveBeenCalledWith(emissionsData);
        expect(setTotal).toHaveBeenCalledWith(getTotal);
        expect(setError).not.toHaveBeenCalled();
    });

    it('should handle error and set default state', async () => {
        const expectedError = new Error('Fetch error');
        const getData = jest.fn().mockRejectedValue(expectedError);

        await fetchData(currentMonth, setData, setTotal, setLoading, setError, getData);

        expect(setData).toHaveBeenCalledWith([]);
        expect(setTotal).toHaveBeenCalledWith(0);
        expect(setLoading).toHaveBeenCalledTimes(2); // should be called twice, once to turn on and once to turn off
        expect(setError).toHaveBeenCalledWith(expectedError);
    });
});

describe('getLabel', () => {
    it('should return the correct label when percentage is greater than 4%', () => {
        const datum = { x: 'Transport', y: 50 };
        const total = 200;
        const expectedLabel = 'Transport\n25%';
        expect(getLabel(datum, total)).toEqual(expectedLabel);
    });

    it('should return null when percentage is less than or equal to 4%', () => {
        const datum = { x: 'Lifestyle', y: 5 };
        const total = 200;
        expect(getLabel(datum, total)).toBeNull();
    });
});

describe('getSelectedLabel', () => {
    test('getSelectedLabel returns the correct label when given valid input', () => {
        const data = [
            { x: 'Label 1', y: 10 },
            { x: 'Label 2', y: 20 },
            { x: 'Label 3', y: 30 }
        ];
        const selectedSlice = 1;

        expect(getSelectedLabel(selectedSlice, data)).toBe('Label 2\n20 lbs CO2');
    });

    test('getSelectedLabel returns null when selectedSlice is null', () => {
        const data = [
            { x: 'Label 1', y: 10 },
            { x: 'Label 2', y: 20 },
            { x: 'Label 3', y: 30 }
        ];
        const selectedSlice = null;

        expect(getSelectedLabel(selectedSlice, data)).toBeNull();
    });

    test('getSelectedLabel returns null when data is empty', () => {
        const data = [];
        const selectedSlice = 0;

        expect(getSelectedLabel(selectedSlice, data)).toBeNull();
    });
});

describe('CategoryChart', () => {
    test('returns the correct label when percentage is greater than 4', () => {
        const datum = { x: 'Transport', y: 100 };
        const total = 1000;
        const label = getLabel(datum, total);
        expect(label).toBe('Transport\n10%');
    });

    test('returns null when percentage is less than or equal to 4', () => {
        const datum = { x: 'Transport', y: 30 };
        const total = 1000;
        const label = getLabel(datum, total);
        expect(label).toBeNull();
    });

    test('calculate total value of all categories', () => {
        const data = [
            { x: "Transport", y: 1000 },
            { x: "Diet", y: 2000 },
            { x: "Home", y: 3000 },
            { x: "Stuff", y: 4000 },
        ];

        const expectedTotal = 10000;
        const total = data.reduce((acc, datum) => acc + datum.y, 0);
        expect(total).toEqual(expectedTotal);
    });

    it('renders correctly', () => {
        const data = [
            { x: "Transport", y: 1000 },
            { x: "Diet", y: 2000 },
            { x: "Home", y: 3000 },
            { x: "Stuff", y: 4000 },
        ];

        const tree = renderer.create(<CategoryChart data={data} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders the "no data" message and "add emissions" button when there is no data', () => {
        const data = [];
        const navigation = {
            navigate: jest.fn(),
        };
        const component = renderer.create(
            <CategoryChart data={data} navigation={navigation} />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

//Checks if daily log renders properly
describe('DailyLog', () => {
    it('renders a bar chart', () => {
        const dataArray = [5, 10, 15, 20, 25];
        const { log } = render(<DailyLog dataArray={dataArray} />);
        expect(log).not.toBeNull();
    });
});