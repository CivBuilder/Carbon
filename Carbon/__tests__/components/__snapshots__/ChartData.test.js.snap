// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryChart renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    You have no data for this month.
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#74C69D",
          "borderRadius": 12,
          "marginTop": 12,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#F7FCF8",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Add Emissions
      </Text>
    </View>
  </View>
</View>
`;

exports[`CategoryChart renders the "no data" message and "add emissions" button when there is no data 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    You have no data for this month.
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#74C69D",
          "borderRadius": 12,
          "marginTop": 12,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#F7FCF8",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Add Emissions
      </Text>
    </View>
  </View>
</View>
`;
